src/enclave/test/http.cpp

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
+#include "../http.h"
+
+#include <doctest/doctest.h>
+#include <queue>
+#include <string>
+#define FMT_HEADER_ONLY
+#include <fmt/format.h>
+
+std::vector<uint8_t> post(const std::string& body)
+{
+  auto req = fmt::format(
+    "POST / HTTP/1.1\r\n"
+    "Content-Type: application/json\r\n"
+    "Content-Length: {}\r\n\r\n{}",
+    body.size(),
+    body);
+  return std::vector<uint8_t>(req.begin(), req.end());
+}
+
+class StubProc : public enclave::http::MsgProcessor
+{
+  std::queue<std::vector<uint8_t>> chunks;
+
+public:
+  virtual void msg(std::vector<uint8_t> m)
+  {
+    chunks.emplace(m);
+  }
+
+  void expect(std::vector<std::string> msgs)
+  {
+    for (auto s : msgs)
+    {
+      if (!chunks.empty())
+      {
+        CHECK(std::string(chunks.front().begin(), chunks.front().end()) == s);
+        chunks.pop();
+      }
+      else
+      {
+        CHECK_MESSAGE(false, fmt::format("Did not contain: {}", s));
+      }
+    }
+    CHECK(chunks.size() == 0);
+  }
+};
+
+TEST_CASE("Complete request")
+{
+  std::string r("{}");
+
+  StubProc sp;
+  enclave::http::Parser p(HTTP_REQUEST, sp);
+
+  auto req = post(r);
+  auto parsed = p.execute(req.data(), req.size());
+
+  sp.expect({r});
+}
+
+TEST_CASE("Parsing error")
+{
+  std::string r("{}");
+
+  StubProc sp;
+  enclave::http::Parser p(HTTP_REQUEST, sp);
+
+  auto req = post(r);
+  req[6] = '\n';
+  CHECK_THROWS_WITH(
+    p.execute(req.data(), req.size()),
+    "HTTP parsing failed: HPE_INVALID_HEADER_TOKEN: invalid character in "
+    "header");
+  sp.expect({});
+}
+
+TEST_CASE("Partial request")
+{
+  std::string r("{}");
+
+  StubProc sp;
+  enclave::http::Parser p(HTTP_REQUEST, sp);
+
+  auto req = post(r);
+  size_t offset = 10;
+
+  auto parsed = p.execute(req.data(), req.size() - offset);
+  CHECK(parsed == req.size() - offset);
+  parsed = p.execute(req.data() + req.size() - offset, offset);
+  CHECK(parsed == offset);
+
+  sp.expect({r});
+}
+
+TEST_CASE("Partial body")
+{
+  std::string r("{\"a_json_key\": \"a_json_value\"}");
+
+  StubProc sp;
+  enclave::http::Parser p(HTTP_REQUEST, sp);
+
+  auto req = post(r);
+  size_t offset = 10;
+
+  auto parsed = p.execute(req.data(), req.size() - offset);
+  CHECK(parsed == req.size() - offset);
+
+  parsed = p.execute(req.data() + req.size() - offset, offset);
+  CHECK(parsed == offset);
+
+  sp.expect({r});
+}
+
+TEST_CASE("Multiple requests")
```

Comment:
Could we add a test case to test the statefulness of the parser (i.e. `IN_MESSAGE`, `DONE`, etc.)?

Reply:
Yes

Reply:
Actually, I'm not sure I agree. As long as we correctly parse and separate the messages, the internal state is of no consequence. If it's incorrect, it will immediately result in incorrect parsin.
It's also going to change quite a bit once we deal with headers, so it's going to be a pain to change for no obvious benefit.